ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Monitor.hs

```diff
@@ -405,13 +469,159 @@ localRoots actions@PeerSelectionActions{ readLocalRootPeers
+-- ledger state judgement value to 'YoungEnough'; and 'hasOnlyBootstrapPeers'
+-- to 'False'.
+--
+monitorBootstrapPeersFlag :: MonadSTM m
+                         => PeerSelectionActions peeraddr peerconn m
+                         -> PeerSelectionState peeraddr peerconn
+                         -> Guarded (STM m) (TimedDecision m peeraddr peerconn)
+monitorBootstrapPeersFlag PeerSelectionActions{ readUseBootstrapPeers
+                                              }
+                          st@PeerSelectionState{ bootstrapPeersFlag
+                                               } =
+  Guarded Nothing $ do
+    ubp <- readUseBootstrapPeers
+    check (ubp /= bootstrapPeersFlag)
+    return $ \_now ->
+      Decision {
+        decisionTrace = [TraceUseBootstrapPeersChanged ubp],
+        decisionJobs  = [],
+        decisionState =
+          st { bootstrapPeersFlag = ubp
+             , ledgerStateJudgement = YoungEnough
+             , hasOnlyBootstrapPeers = False
+             }
+      }
+
+-- | Monitor 'LedgerStateJudgement', if it changes, depending on the value we
+-- just need to update 'PeerSelectionTargets'. If the ledger state changed to
+-- 'TooOld' we set all other targets to 0 and wait the governor waits for all
+-- active connections to drop and then set the targets for bootstrap peers.
+-- However if the state changes to 'YoungEnough' we reset the targets back to
+-- their original values.
+--
+-- It should be noted if the node has bootstrap peers disabled then this
+-- monitoring action will be disabled.
+--
+monitorLedgerStateJudgement :: MonadSTM m
+                            => PeerSelectionActions peeraddr peerconn m
+                            -> PeerSelectionState peeraddr peerconn
+                            -> Guarded (STM m) (TimedDecision m peeraddr peerconn)
+monitorLedgerStateJudgement PeerSelectionActions{ readLedgerStateJudgement
+                                                }
+                            st@PeerSelectionState{ bootstrapPeersFlag
+                                                 , ledgerStateJudgement
+                                                 }
+  | isBootstrapPeersEnabled bootstrapPeersFlag =
+    Guarded Nothing $ do
+      lsj <- readLedgerStateJudgement
+      check (lsj /= ledgerStateJudgement)
+      st' <- case lsj of
+        TooOld      ->
+          return st
+            { ledgerStateJudgement = lsj
+            , targets =
+                PeerSelectionTargets
+                  { targetNumberOfRootPeers                 = 0
+                  , targetNumberOfKnownPeers                = 0
+                  , targetNumberOfEstablishedPeers          = 0
+                  , targetNumberOfActivePeers               = 0
+                  , targetNumberOfKnownBigLedgerPeers       = 0
+                  , targetNumberOfEstablishedBigLedgerPeers = 0
+                  , targetNumberOfActiveBigLedgerPeers      = 0
+                  }
```

Comment:
I see that the `targetPeers` is using `isNodeAbleToMakeProgress`, which makes sure that when churns updates the `targets` they are done when `isNodeAbleToMakeProgress` returns `True`.

I think this still can lead to problems.  Churn is updating the targets in non atomic way, i.e. in a sequence of atomic operations done in different intervals.  If this job intervenes between one of them, it could confuse churn.   It probably didn't show up in the tests because we don't have a test that churn is able to set the targets to the right values which we need to add.

I think it was a right step to disable churn in bootstrap phase (from the moment we enter it, until we finished transitioning out of it).

Could you find out how churn interacts with local roots.  I don't have a clear understanding how `{hot,warm}Valency` interacts with churn (it could be that churn is used to swap some `hot`, `warm` and `cold` local roots).

Reply:
`targetPeers` is only disabled when the node is put in `TooOld` state, this state the node can not make any progress and should focus in demoting all peers. After successfully cleaning the state the node can make progress again and the targets will be restored and the `targetPeers` monitoring action will work again.

IIRC churn only changes the targets for hot/warm/cold and disregards local roots, it leaves it to the outbound governor to make sure to reconnect to them, but I will double check

Reply:
We discussed this in slack.  The key points for me are that:
* we only disable `targetPeers` monitoring job (and thus churn) when we do one of the transitions to or from bootstrap peers;
* targets are set by churn, which includes targets in bulk sync & deadline mode.  In bootstrap phase the bulk sync targets will be applied, so we won't be connected to too many bootstrap peers;
* churn in bootstrap phase will help to find the closest bootstrap peers (as now in the bulk sync mode).